File: arrays/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1:
            return [nums]

        res = []

        for i in range len(nums):
            num = nums[i]

            remaining = nums[:i] + nums[i + 1:]

            for perm in self.permute(remaining):
                res.append([num] + perm)

        return res

File: arrays/permutations.rb
def permute(nums)
    return [nums] if nums.size == 1
    res = []

    nums.each_with_index do |num, ind|

        remaining = nums.dup
        remaining.delete_at(ind)

        permute(remaining).each do |perm|
            res << [num] + perm
        end
    end

    res
end